src/providers/kuksa_val_v1/provider.rs

```diff
@@ -89,23 +98,45 @@ impl ProviderInterface for Provider {
         signal_data: &[Signal],
         iteration: u64,
     ) -> Result<Instant, PublishError> {
-        let updates = Vec::from_iter(signal_data.iter().map(|signal| {
-            proto::EntryUpdate {
-                entry: Some(proto::DataEntry {
-                    path: signal.path.clone(),
-                    value: Some(proto::Datapoint {
-                        timestamp: None,
-                        value: Some(
-                            n_to_value(self.data_types_map.get(&signal.path).unwrap(), iteration)
-                                .unwrap(),
-                        ),
+        let updates = if iteration == 0 {
+            Vec::from_iter(signal_data.iter().map(|signal| {
+                let metadata = self.metadata.get(&signal.path).unwrap();
+                let mut new_value = n_to_value(metadata, iteration).unwrap();
+                if let Some(value) = self.initial_signals_values.get(&signal.path) {
+                    if DataValue::from(&Some(new_value.clone())) == *value {
+                        new_value = n_to_value(metadata, iteration + 1).unwrap();
+                    }
+                }
+                proto::EntryUpdate {
+                    entry: Some(proto::DataEntry {
+                        path: signal.path.clone(),
+                        value: Some(proto::Datapoint {
+                            timestamp: None,
+                            value: Some(new_value),
+                        }),
+                        actuator_target: None,
+                        metadata: None,
                     }),
-                    actuator_target: None,
-                    metadata: None,
-                }),
-                fields: vec![proto::Field::Value as i32],
-            }
-        }));
+                    fields: vec![proto::Field::Value.into()],
+                }
+            }))
+        } else {
```

Comment:
Why should we store the initial values in the provider(s) and complicate the publish function like this (or all of them)?

It's enough to change the iteration number outside of the `publish` function based on the initial value. The provider doesn't need to know about this at all.

Reply:
I know it makes thing more complicated, however the function responsible for generating signal values `n_to_value` is API specific and its implementation is not known at `publish` level.
I can not image a way to generate a signal value where `publish` get called in a generic way for each API, it could be done but maybe would be more complicated than just make the `publish` method more complex